stl/inc/xstring

```diff
@@ -4332,13 +4340,15 @@ _NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
     return _STD move(_Left.append(_Right));
 }
 
+
 template <class _Elem, class _Traits, class _Alloc>
 _NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
     basic_string<_Elem, _Traits, _Alloc>&& _Left, basic_string<_Elem, _Traits, _Alloc>&& _Right) {
-    if (_Right.size() <= _Left.capacity() - _Left.size() || _Right.capacity() - _Right.size() < _Left.size()) {
-        return _STD move(_Left.append(_Right));
-    } else {
+    bool _Equal_allocs = allocator_traits<_Alloc>::is_always_equal::value || _Left._Getal() == _Right._Getal();

```

Comment:
`std::allocator_traits::is_always_equal` is implemeted unconditionally, although appered in C++17. Does it mean I don't need to check for C++17 here?

Reply:
Prior to C++17 user's `std::allocator_traits` specializations are not required to have `is_always_equal` so you have to either check for C++17 or (what I think is better option) check if `is_always_equal` is present and if not probably assume it'd be `false_type` or use `is_empty`.

Reply:
`allocator_traits` itself already does the `is_always_equal` lazy checking.

Reply:
But user can provide his own specialization of `allocator_traits`. At least standard permits it.

Reply:
The standard permits it, but we generally don't care, because users don't do that. The standard permits the user providing their own `std::vector` too.